### PR TITLE
Image Task

### PR DESCRIPTION
**Pull request by tomyyyD**
Image task runs every 5 seconds. If 60 seconds have elapsed since the camera was last turned on, it will turn on the camera and initiate getting an image. If the camera is on but hasn't been contacted, the task will check for a byte that the camera will send when it starts up to confirm the connection. If the camera is on and has been contacted, it will take an image, store the image on its SD card and send that image over in 500B packets to the mainboard. Each time the task runs, it will spend a maximum of 2 seconds sending as many packets as it can. When the image is finished the camera will be turned off.

### CODE REVIEW

**Comment by tomyyyD** [comment]
closes #368

File: state_machine/applications/flight/Tasks/image_task.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
+# Transmit "Hello World" beacon
```
Comment (thetazero) [inline state_machine/applications/flight/Tasks/image_task.py:1]: comment here probably wasn't intended

**Review by thetazero — CHANGES_REQUESTED** [review]
This looks really good, I'm really happy with the design you've chosen. 

Two things to consider (that don't necessarily have to be done in this pr)
1. As it stands this is very hard to run automatic tests on. 
We could improve this by separating the communication protocol, the communication drivers, and the actual task further. 
2. For our GNC code a 2 second outage is really annoying to deal with. We can fix this by making it significantly shorter, or by making the code that loads an image over uart not impose significant outages (I think this is nicer, because then we can load a full image per call)

File: state_machine/applications/flight/Tasks/image_task.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+IMAGE_START = 0xAC
+IMAGE_MID = 0xAD
+IMAGE_END = 0xAE
+IMAGE_CONF = 0xAF
+PACKET_REQ = 0xB0
+NO_IMAGE = 0xB1
+MAX_CONF_ATTEMPTS = 3
+
+image_buffer = bytearray(499)
+header_buffer  = bytearray(1)
```
[the remark this answers is not in the document]
Comment (tomyyyD) [reply state_machine/applications/flight/Tasks/image_task.py]: moved all buffer interactions into a separate camera driver, so hopefully that makes things clearer. I also added comments for what each buffer does.

File: state_machine/applications/flight/Tasks/image_task.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                    header_buffer[0] = CONFIRMATION_RECEIVE_CODE
+                    cubesat.uart.write(header_buffer)
+                    self.debug("responding", level=2)
+                    cubesat.uart.reset_input_buffer()
+                    self.cam_active = True
+                    self.conf_attempts = 0
+                    break
+            self.conf_attempts += 1
+            if self.conf_attempts > MAX_CONF_ATTEMPTS:
+                # turn off camera to not waste battery as something is wrong
```
[the remark this answers is not in the document]
Comment (tomyyyD) [reply state_machine/applications/flight/Tasks/image_task.py]: created a few alerts that represent the image task and camera device. camera_available, camera_failed, image_queue_full.

File: state_machine/applications/flight/Tasks/image_task.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        else:
+            self.debug("camera asleep")
+            if not iq.empty():
+                self.debug(f"{iq.peek()}", level=2)
+            else:
+                self.debug("queue empty", level=2)
+
+        collect()
+
+    async def get_packets(self, st):
```
[the remark this answers is not in the document]
Comment (tomyyyD) [reply state_machine/applications/flight/Tasks/image_task.py]: I'm not entirely familiar with how we are deferring execution. Does making the function async and awaiting its execution not defer execution?

File: state_machine/applications/flight/Tasks/image_task.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        done = False
+        while monotonic() - 2 < st and not done:
+            done = self.get_packet()
+
+    def get_packet(self) -> bool:
```
[the remark this answers is not in the document]
Comment (tomyyyD) [reply state_machine/applications/flight/Tasks/image_task.py]: I created a separate camera driver that has the get_packet function

**Comment by tomyyyD** [comment]
If anyone has thoughts on a better way for the camera driver's `get_packet()` function to return the status of the packet, let me know I'm open to ideas.

File: applications/flight/Tasks/image.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            self.last_time = monotonic()
+            return True
+        return False
+
+    async def main_task(self):
+        """
+        gets an image from the camera board
+        """
+        if not cubesat.camera:
+            alerts.clear(self.debug, 'camera_available')
```
Comment (thetazero) [inline applications/flight/Tasks/image.py:45]: imo we can skip this, the hardware monitor will get to it.

Comment (thetazero) [reply applications/flight/Tasks/image.py:45]: maybe a camera_task alert

File: applications/flight/Tasks/hw_monitor.py [hunk 7]
```diff
@@ -15,3 +15,4 @@ async def main_task(self):
         alerts.set_value(self.debug, 'radio_available', cubesat.radio is not None)
         alerts.set_value(self.debug, 'rtc_available', cubesat.rtc is not None)
         alerts.set_value(self.debug, 'neopixel_available', cubesat.neopixel is not None)
+        alerts.set_value(self.debug, 'camera_available', cubesat.camera is not None and cubesat.sdcard is not None)
```
Comment (thetazero) [inline applications/flight/Tasks/hw_monitor.py:18]: probably want to split this into a camera/std card alert. This was we'd know which devices are actually offline (unless we genuinely cannot tell if the camera is available when the sdcard is not)

File: drivers/pycubedmini/lib/camera.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        working file.
+
+        returns packet and a code which correspondes to what happened when getting the packet
+        - 0: success
+        - 1: success, image not interesting
+        - 2: success, first packet
+        - 3: success, last packet
+        - 4: camera must not have received our confirmation byte as we are still getting
+             the confirmation send code, send confirmation byte again
+        - 5: failed to get packet
```
Comment (thetazero) [inline drivers/pycubedmini/lib/camera.py:57]: I think it'd be more readable if we return constants like `SUCCESS`, `SUCESS_NOT_INTERESTING_IMAGE`, etc... flags are a little difficult to read. Then on the application side we could just compare to these constants.

**Review by thetazero — CHANGES_REQUESTED** [review]
Still a little concerned about not having testing for this. 
I think it's fine to merge without given that it's a big effort to bring tests like these up, but lets discuss a plan first to make sure this code is possible to test.

**Comment by tomyyyD** [comment]
closes #392

**Review by thetazero — APPROVED** [review]
Looks good

**Pull request merged**